### ISSUE TITLE
How to connect an external experiment with goals in alephbet?

### ISSUE THREAD

**Issue by tttp** [Alephbet/alephbet#3]
Hi,

I'm going to A/B test my mailing too. The "google analytics" way of tracking that is to use utm_xxx parameters in the url, so http://example.org?utm_campaign=ab_test_42&utm_source=newsletter_june&utm_medium=mail&utm_content=red-variant
https://support.google.com/analytics/answer/1033867?hl=en

Is this possible to either re-use the same utm_xxx param or add some alep_experiment=ab_test_42&alep_variant=red-variant?

**Comment by tttp** [comment]
I'm fully aware that given they aren't a ton of analysis yet on the results as stored in google analytics, so trying to generate the same events without doing the a/b in alephbet isn't necessarily super useful yet, but I'd like to store my ab tests results the same way to reuse the yet to come analysis tool...

**Comment by gingerlime** [comment]
I think you're looking at forcing a variant. I haven't implemented it yet... see https://github.com/gingerlime/alephbet/blob/master/alephbet.js.coffee#L50. It shouldn't be too difficult though, but I wasn't sure about the best approach for fetching params from the URL. Cohorts.js implemented this [here](https://github.com/jamesyu/cohorts/blob/master/lib/cohorts.js#L57-L68).

I'd like to use some external library like URI.js perhaps, but I also need to check how much weight it might add to the minified file (keeping size small is quite important). Writing something similar to cohorts isn't that difficult either, but I want something slightly more elegant. I just didn't get round to it. 

As a quick hack, you could do something like this (untested):

```javascript
var experiment = // ... get your experiment name from the utm tag or param
var variant = // ... same
var my_experiment = new AlephBet.Experiment({ // ...
  // make sure we don't trigger the experiment yet
  trigger: function() { return false; }
});
// forcing variant by writing it into storage
my_experiment.storage().set(experiment + ':variant', variant);
// overwriting trigger
my_experiment.options.trigger = function() { return true; };
// running the experiment
my_experiment.run();
```

**Comment by tttp** [comment]
Aaah, interesting idea and like it, but that wasn't what I had in mind.

Let me try to clarify: what I want to test here is the content of the email, so the page isn't going to be different, just the content of the email (that is obviously not going to be generated by alephbet

however, I'd like to track goals using the lib:
- how many pages are viewed (opening rate)
- how many click on the button (or fill form, or whatever action)

anyway, so the point is **not** to register any event when the experiment "runs" (but isn't necessarily armful either, it's just measuring opening rate,  so a goal, not real experiments, that are done elsewhere

    var experiment = // ... get your experiment name from the utm tag or param
    var variant = // ... same
    var my_experiment = new AlephBet.Experiment({ // ...
      // make sure we don't trigger the experiment yet
      trigger: function() { return false; }
      //the variants do nothing, ideally, no need to defines them at all
      variants: {"mail A": {activate:function() {return true}}
        {"mail B": {activate:function() {return true}}
    });
    // forcing variant by writing it into storage
    my_experiment.storage().set(experiment + ':variant', variant);
    // overwriting onInitialize, probably not the right syntax...
    my_experiment.onInitialize = function () { return true};
    my_experiment.options.trigger = function() { return true; };
    // running the experiment
    my_experiment.run();

    AlephBet.Goal('page view', {unique: false});
    AlephBet.Goal('form submit');
    ...

**Comment by gingerlime** [comment]
sounds to me like you still have to force a variant based on the utm param (or other param), which indicates which email led people to the page(s) you're tracking. No? Otherwise, the goal will be attributed randomly to one variant, but not necessarily the one matching the email version.

If I understand things: You generate two types of emails: email A or email B. Send them randomly to people (roughly 50% of the people will receive A, the other 50% B). email A has a link with a param experiment=my_experiment & variant=A. email B has a link with param experiment=my_experiment & variant=B.

When someone clicks on the link in email A - they go to a page, passing those params. You then take those params as I explained, and *force* the variant A. Same goes with B, you force B. Otherwise, the variant will be randomly picked irrespective of the email...

Does it make sense?

**Comment by gingerlime** [comment]
... and yes - the variant function isn't important. it can simply return nothing. AlephBet doesn't really care what you do when you activate a variant.

**Comment by tttp** [comment]
Correct and your code makes sense, my additional points were that.
- I need to force register the variant (to be used in the goal) indeed (cf your code)
- I _would like_ to skip the onInitialize event
- I _would like_ to avoid re-defining the variants on the Experiment, to avoid having to sync the variants name between my mailing service that does the ab testing of the mails and the landing page(s)

**Comment by gingerlime** [comment]
> I _would like_ to skip the onInitialize event

personally, I would keep it, but if you do want to skip it, then I would use a custom tracking adapter that does nothing when `onInitialize` gets called and pass this custom tracking adapter to the experiment

> I _would like_ to avoid re-defining the variants on the Experiment, to avoid having to sync the variants name between my mailing service that does the ab testing of the mails and the landing page(s)

That should be possible, for example, if you pass all variants to the page. Something like:

```javascript
var variants_arr = // ... get the variants as an array from the URL params (so it will contain ['email_A', 'email_B'])

// creating an object with each variant, so you can pass this to AlephBet
var variants = {};
(variants_arr).forEach(function(v) {
  variants[v] = {
    activate: function() {}
  }
});

var my_experiment = new AlephBet.Experiment({
  variants: variants,
  tracking_adapter: // pass your custom tracking adapter that doesn't do anything with onInitialize
  // ...
});
```

Does this sound like a workable solution?

**Comment by gingerlime** [comment]
Just a quick note, I released 0.5.0 and the internal API changed a bit, so now `onInitialize` is called `experiment_start`, and `onEvent` is called `goal_complete`. See the latest README for details. I also created a screencast, which might explain a few things better than the README.

**Comment by gingerlime** [comment]
I'm closing this for now since it wasn't updated in a while, but feel free to re-open if you wish to discuss this further.
